Add new quest to determine the maxheight

ENT8R opened the pull request:
This PR is a little bit more extensive... It adds a quest for determining the maximum height of e.g. tunnels or covered ways, parking entrances for underground and multi-storey parkings and of course height restrictors. It fixes #399, #421, #447.
Sorry for creating so many PRs in the last time, but I hope this helps to get the issue amount below 100 😄

Here are some things I have questions about:

- [ ] Can somebody maybe proofread the [list of height units per country](https://github.com/ENT8R/StreetComplete/blob/maxheight/res/country_metadata/heightUnit.yml)?
- [ ] If there is no sign, should we tag it as `maxheight=default` (used about 10K times) or `maxheight=no_sign` as proposed by the wiki but only used 14 times according to taginfo...
see [the source code](https://github.com/ENT8R/StreetComplete/blob/59b437425c51bacc9c0fd8c8b9866c86e4f474e2/app/src/main/java/de/westnordost/streetcomplete/quests/max_height/AddMaxHeight.java#L66) for more information
- [ ] Should [the query](https://github.com/ENT8R/StreetComplete/blob/59b437425c51bacc9c0fd8c8b9866c86e4f474e2/app/src/main/java/de/westnordost/streetcomplete/quests/max_height/AddMaxHeight.java#L47) be improved to add [the same functionality](https://github.com/mmd-osm/osm-maxheight-map/blob/1cd999d3d276bcef04224ccf045e0e81ec051a42/js/map.js#L282) as the maxheight map by @mmd-osm does?
- [ ] If the unit is metric but the input value is an integer (without values after the comma 😄), should automatically a trailing `.0` be added by SC? This is also the current behaviour...

The quest looks like this in countries with the **metric system** known:
<img src="https://user-images.githubusercontent.com/25306497/37171693-fba71d4c-230e-11e8-9709-81e11fe9b0a2.png" width="300px"> <img src="https://user-images.githubusercontent.com/25306497/37171722-16dd6c6a-230f-11e8-9cad-060b0a058330.png" width="300px">

<img src="https://user-images.githubusercontent.com/25306497/37171718-14bb4a42-230f-11e8-8dd1-1aedfe64ec26.png" width="300px">

And this is how the quest will look like in countries with the **imperial system**

<img src="https://user-images.githubusercontent.com/25306497/37171738-24b1d394-230f-11e8-9654-8db5bd89e5e6.png" width="300px"> <img src="https://user-images.githubusercontent.com/25306497/37171741-272228ae-230f-11e8-8d52-df70cab84783.png" width="300px">

matkoniecz commented:
I was just reading code, but 

> This height looks implausible

(if necessary) should be rather something like

> This height looks implausible, are you sure that it is correct?

westnordost commented:
Wow, this is a complex one. You even took care of those imperial countries, nice.

> Sorry for creating so many PRs in the last time, but I hope this helps to get the issue amount below 100 😄

It would be better if you do one thing after another and see it to the end, rather than creating one construction site after another, in my opinion. These days, it feels like I am spending 90% of the time I have for SC on commenting on tickets or reviewing PRs. 
On that note, I feel sorry for fritruc that I will not help him nor review his PR (in this state), seeing the amount of code he fabricated, but I just don't have the time for that. Perhaps you could give him a hand and point out the blunders and possible improvements in his code since you know quite a bit of Android development, the SC source code by now and the quality I demand?

Anyway, will review your code now...

westnordost commented on app/src/main/java/de/westnordost/streetcomplete/quests/max_height/AddMaxHeight.java:
```diff
@@ -0,0 +1,99 @@
+
+	@Override public int getTitle() { return R.string.quest_maxheight_title; }
+	@Override public int getDefaultDisabledMessage() { return 0; }
+	@Nullable @Override public Boolean isApplicableTo(Element element) { return null; }
```
I think it is. But it would be a long chain of if-checks, I figure. You could use several tagfilters. I.e.... (pseudo code)

```
TagFilterExpression nodeFilter = parser.parse("nodes with (barrier=height_restrictor or amenity=parking_entrance and parking ~ underground|multi-storey) and !maxheight and !maxheight:physical");
TagFilterExpression wayFilter = parser.parse("ways with highway~(primary|secondary|tertiary|trunk)(_link)?|motorway|service|residential|unclassified|living_street and (covered=yes or tunnel~yes|building_passage) !maxheight and !maxheight:physical");
return nodeFilter.matches(element) || wayFilter.matches(element);
```

To be honest, with a little extension on the TagFilterExpression class (adding a method that outputs the TagFilterExpression without the full thing (bbox and out geom meta;) as Overpass QL, perhaps these filters could be used as well in the getOverpassQuery method.

westnordost commented on app/src/main/java/de/westnordost/streetcomplete/quests/max_height/AddMaxHeightForm.java:
```diff
@@ -0,0 +1,196 @@
+
+	private int getMaxHeightLayoutResourceId()
+	{
+		return isMetric ? R.layout.quest_max_height : R.layout.quest_max_height_us;
```
Unfortunately, the US is not the only country that uses imperial units. You need to separate the selection of the layout from the selection of which unit to select. [I.e. this is how the signs look like in UK apparently.](https://en.wikipedia.org/wiki/File:UK_traffic_sign_629.2A.svg)

1. Traffic sign layout: most countries follow the [International Standard](https://en.wikipedia.org/wiki/Vienna_Convention_on_Road_Signs_and_Signals), see also the [comparison of European road signs](https://en.wikipedia.org/wiki/Comparison_of_European_road_signs). This is the layout I saw in your screenshot. However, I think you should also include the "**m**" in the form, even if it is not a dropdown menu, because it is an element in **all** the signs.
Some countries follow the [US standard (MUTCD)](https://en.wikipedia.org/wiki/Comparison_of_MUTCD-influenced_traffic_signs). In the linked page, search for "Height restriction ahead" to see how the sign look there. So, the layout should look [like that](https://en.wikipedia.org/wiki/Comparison_of_MUTCD-influenced_traffic_signs#/media/File:Brasil_A-37.svg). If you feel extra-cool, you could make extra forms even on variations, i.e. for Australia.

2. Meters vs Feet+Inch: So, both layouts need to have both input fields. These input fields can simply be made visible/invisible depending on the country or user's selection where both systems may be used.

westnordost commented on app/src/main/java/de/westnordost/streetcomplete/quests/max_height/AddMaxHeightForm.java:
```diff
@@ -0,0 +1,196 @@
+					isMetric = true;
+					setMaxHeightSignLayout(getMaxHeightLayoutResourceId());
+				});
+			}
```
Okay, that is also a possible solution. Why though did you not solve it in the same way as in the maxspeed quest though? To me, it looks like the swifter solution.

ENT8R replied:
I had some trouble with displaying the spinner on my API 19 emulator which I could not yet resolve... For the max speed quest everything worked fine but for this quest it always gave me this output: `Could not find class 'android.widget.ThemedSpinnerAdapter', referenced from method android.support.v7.widget.AppCompatSpinner$DropDownAdapter.<init>`
See also https://stackoverflow.com/questions/34014843/could-not-find-class-android-widget-themedspinneradapter-android-studio/37050929 which does not have a proper solution...

westnordost replied:
Does the same problem exist for the maxspeed quest? (I.e. go to Myanmar and download the quest there, the spinner should show)

ENT8R replied:
As I said: For the max speed quest, everything worked fine (I just tested it again...)

westnordost commented on app/src/main/java/de/westnordost/streetcomplete/quests/max_height/AddMaxHeightForm.java:
```diff
@@ -0,0 +1,196 @@
+		if (isMetric)
+		{
+			/*TODO: should a trailing zero be added or does it even work without?
+			https://taginfo.openstreetmap.org/keys/maxheight#values contains both...*/
```
Why should it? I guess in doubt, use whatever is clearly favoured by the community. If nothing is really favoured, I would go with not adding extra-work here (not adding .0)

ENT8R replied:
The [examples of the wiki](https://wiki.openstreetmap.org/wiki/Key:maxheight) do also only list values without a trailing .0, so I will remove this.

goldfndr commented on app/src/main/java/de/westnordost/streetcomplete/quests/max_height/AddMaxHeight.java:
```diff
@@ -0,0 +1,99 @@
+			"(" +
+			" node['barrier'='height_restrictor']" +  QUERY_RESTRICTIONS + ";" +
+			" node['amenity'='parking_entrance']['parking'~'^(underground|multi-storey)$']" +  QUERY_RESTRICTIONS + ";" +
+			" way['highway'~'^(" + ROADS + ")$']['tunnel'~'^(yes|building_passage)$']" +  QUERY_RESTRICTIONS + ";" +
```
Please add `avalanche_protector` to tunnel types.

westnordost commented on app/src/main/java/de/westnordost/streetcomplete/quests/max_height/AddMaxHeightForm.java:
```diff
@@ -0,0 +1,196 @@
+			else
+			{
+				return null;
+			}
```
you can replace this with simply return `null`. However, if this method is able to return `null`, you need to make sure that any point where this is used, a nullcheck is made.
It might be easier (=less code) to simply return an empty string if the text field is empty and only check for that where we are really interested in that (when checking if there had been changes).

ENT8R replied:
I don't think that nullchecks are needed for this as this method is only used after `onClickOk()` with the first method checking wether there are changes... And if not, there is only a toast and no more code gets executed...

westnordost commented on app/src/main/java/de/westnordost/streetcomplete/quests/max_height/AddMaxHeightForm.java:
```diff
@@ -0,0 +1,196 @@
+			if (!heightInput.getText().toString().isEmpty())
+			{
+
+				return heightInput.getText().toString().replace(",", ".");
```
Where does the "," come from?

rugk replied:
Ah, I see. If the user enters it with , instead of . (in Germany, e.g.) However, I think, even the German keyboard layout for numbers limits the input to `.` as a decimal separator.

ENT8R replied:
Right, this is why I added it...

> I think, even the German keyboard layout for numbers limits the input to . as a decimal separator.

I don't think so... I even specified the possible input values (the same is done for the housenumber quest too) here:
https://github.com/ENT8R/StreetComplete/blob/59b437425c51bacc9c0fd8c8b9866c86e4f474e2/app/src/main/res/layout/quest_max_height.xml#L22

westnordost replied:
ENT8R, the line you cited is not there to allow German decimal points but to allow an input like "3,5" (multiple housenumbers).

I do not believe that Android multilingual support is supposed to be that crappy that it leaves it to the app to "understand" the number of the user input. 

You should set `android:inputType=numberDecimal` and probably remove the android:digits="0123456789.,", then everything should be fine.

ENT8R replied:
Alright. Will do that!

westnordost commented on app/src/main/java/de/westnordost/streetcomplete/quests/max_height/AddMaxHeightForm.java:
```diff
@@ -0,0 +1,196 @@
+		{
+			if (!feetInput.getText().toString().isEmpty() && !inchInput.getText().toString().isEmpty())
+			{
+				return feetInput.getText().toString() + "." + inchInput.getText().toString();
```
Not sure why you do this, but I find this dangerous. Ten inch are not one foot, twelve inch are one foot. By introducing the dot, you seem to imply that the inch are the decimal, which they are not.

ENT8R replied:
Yeah, true... I will have to find a better solution...

westnordost commented on app/src/main/res/drawable/background_max_height_sign.xml:
```diff
@@ -0,0 +1,17 @@
+    <path
+        android:fillColor="#fff"
+        android:pathData="M17.967,22.201c3.207,-1.6 5.235,-4.876 5.236,-8.461 -0.002,-3.585 -2.029,-6.86 -5.236,-8.461l-4.224,3.698 -4.217,-3.694C6.317,6.881 4.287,10.156 4.283,13.74c0.005,3.585 2.035,6.859 5.244,8.456l4.217,-3.694z"/>
+</vector>
```
Actually, it would also be possible to use the simple layer-list drawable (see `background_maxspeed_sign.xml`). The advantage would be that it is also displayed as vector on older Android versions.

ENT8R replied:
Is there some chance to convert it? I don't think so but I'm not familiar at all with those layer-list drawables...

westnordost replied:
No. But hmm, don't mind about that. Turns out that those two triangles are not possible with shape drawables alone.

ENT8R replied:
Alright.

westnordost commented on app/src/main/res/values/strings.xml:
```diff
@@ -495,4 +495,15 @@ Otherwise, you can download another keyboard in the app store. Popular keyboards
+    <string name="quest_maxheight_tunnel_title">What is the height limit of this tunnel?</string>
+    <string name="quest_maxheight_answer_noSign">There is no sign</string>
+    <string name="quest_maxheight_answer_noSign_confirmation_title">Are you sure there is no sign?</string>
+    <string name="quest_maxheight_answer_noSign_confirmation">Did you check at both ends? If there is no sign, default limits apply.</string>
```
This doesn't make sense. What are the "default limits"? Also, it doesn't make sense (IMO) to ask the user to look at "both ends". I think it doesn't matter from which side you enter a tunnel for the height restriction to apply. :-D

goldfndr commented on app/src/main/java/de/westnordost/streetcomplete/quests/max_height/AddMaxHeightForm.java:
```diff
@@ -0,0 +1,196 @@
+	{
+		double height = Double.parseDouble(getHeightFromInput());
+		double heightInMeter = isMetric ? height : feetToMeter(height);
+		return heightInMeter > 6;
```
Low heights might also be implausible, maybe > 2 m?

westnordost commented on res/country_metadata/heightUnit.yml:
```diff
@@ -0,0 +1,13 @@
+# https://en.wikipedia.org/wiki/Metrication#Overview
+# https://en.wikipedia.org/wiki/Imperial_units#Current_use
+default: [m]
+CA: [ft, m]
```
Canada should be meter only, see https://en.wikipedia.org/wiki/Comparison_of_MUTCD-influenced_traffic_signs
Or, at least, m first, then ft.

goldfndr commented on app/src/main/java/de/westnordost/streetcomplete/quests/max_height/AddMaxHeightForm.java:
```diff
@@ -0,0 +1,196 @@
+		{
+			if (!feetInput.getText().toString().isEmpty() && !inchInput.getText().toString().isEmpty())
+			{
+				return feetInput.getText().toString() + "." + inchInput.getText().toString();
```
Although simplifying inches to a range from 12 to 10 doesn't seem so bad, an 11 inch measurement will give an odd result. (I admittedly didn't check against boundary condition.)

westnordost commented:
Hmm, not sure why it is not shown (to me), but I also wrote:

>Unfortunately, the US is not the only country that uses imperial units. You need to separate the selection of the layout from the selection of which unit to select. I.e. this is how the signs look like in UK apparently.
> 1. Traffic sign layout: most countries follow the International Standard, see also the comparison of European road signs. This is the layout I saw in your screenshot. However, I think you should also include the "m" in the form, even if it is not a dropdown menu, because it is an element in all the signs.
    Some countries follow the US standard (MUTCD). In the linked page, search for "Height restriction ahead" to see how the sign look there. So, the layout should look like that. If you feel extra-cool, you could make extra forms even on variations, i.e. for Australia.
> 2. Meters vs Feet+Inch: So, both layouts need to have both input fields. These input fields can simply be made visible/invisible depending on the country or user's selection where both systems may be used.

westnordost commented:
> Should the query be improved to add the same functionality as the maxheight map by @mmd-osm does?

No, because that would potentially show very long stretches of road which would need to be split up. So, this would introduce possibly wrong data because the height restriction surely is not valid for the whole road/stretch but only for the area below the bridge.

mmd-osm commented on app/src/main/java/de/westnordost/streetcomplete/quests/max_height/AddMaxHeight.java:
```diff
@@ -0,0 +1,99 @@
+			/*TODO: maxheight=default or maxheight=no_sign?
+			according to https://taginfo.openstreetmap.org/keys/maxheight#values maxheight=no_sign is only used 14 times,
+			while maxheight=default is used about 10K times*/
+			changes.add("maxheight", "default");
```
They’re documented on the Osm wiki

matkoniecz replied:
> What is the information gain here?

I guess that in typical country it is mandatory to sign maxheight if it is below some value (I would expect it to be chosen so normal-sized vehicles can pass). So lack of sign indicates that normal vehicles can pass trough. Also, with this it is always possible to tag something, making this quest possible.

> about an estimate of the height and tag the estimate then

At least I am not really able to estimate height with any kind of accuracy that would make worth tagging it.

mmd-osm commented:
> westnordost commented 7 hours ago
Should the query be improved to add the same functionality as the maxheight map by @mmd-osm does?
No, because that would potentially show very long stretches of road which would need to be split up. So, this would introduce possibly wrong data because the height restriction surely is not valid for the whole road/stretch but only for the area below the bridge.

BTW: Some queries in maxheight map depend on extensive post processing, using the query results as-is will not suffice in all cases.

westnordost commented:
The estimate dialog could include the hint to take your own body height as a measure and estimate how many times you'd fit stacked on top of each other. Would that help you? As said, it should be an estimate only and clearly be tagged as such.

On 9 March 2018 09:09:36 CET, Mateusz Konieczny <notifications@github.com> wrote:
>matkoniecz commented on this pull request.
>
>
>
>> +
>+	@Override public AbstractQuestAnswerFragment createForm()
>+	{
>+		return new AddMaxHeightForm();
>+	}
>+
>+	@Override public void applyAnswerTo(Bundle answer,
>StringMapChangesBuilder changes)
>+	{
>+		boolean hasNoSign = answer.getBoolean(AddMaxHeightForm.NO_SIGN);
>+		String maxheight = answer.getString(AddMaxHeightForm.MAX_HEIGHT);
>+		if(hasNoSign)
>+		{
>+			/*TODO: maxheight=default or maxheight=no_sign?
>+			according to
>https://taginfo.openstreetmap.org/keys/maxheight#values
>maxheight=no_sign is only used 14 times,
>+			while maxheight=default is used about 10K times*/
>+			changes.add("maxheight", "default");
>
>> What is the information gain here?
>
>I guess that in typical country it is mandatory to sign maxheight if it
>is below some value (I would expect it to be chosen so normal-sized
>vehicles can pass). So lack of sign indicates that normal vehicles can
>pass trough. Also, with this it is always possible to tag something,
>making this quest possible.
>
>> about an estimate of the height and tag the estimate then
>
>At least I am not really able to estimate height with any kind of
>accuracy that would make worth tagging it.
>
>-- 
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/westnordost/StreetComplete/pull/960#discussion_r173386312

-- 
Diese Nachricht wurde von meinem Android-Mobiltelefon mit K-9 Mail gesendet.

mmd-osm commented:
ping //cc: @slhh

mmd-osm commented on app/src/main/java/de/westnordost/streetcomplete/quests/max_height/AddMaxHeight.java:
```diff
@@ -0,0 +1,99 @@
+
+	/*the query does not exclude elements which are not accessible by the public,
+	because a maxheight sign will very probably be visible at the entrance or beginning*/
+	private static final String QUERY_RESTRICTIONS = "['maxheight'!~'.']['maxheight:physical'!~'.']";
```
You can use `!tag` instead of `"tag"!~'.'` here : e.g. `[!maxheight]`, or `[!"maxheight:physical"]`

slhh commented:
> If there is no sign, should we tag it as maxheight=default (used about 10K times) or maxheight=no_sign as proposed by the wiki but only used 14 times according to taginfo...

If there is no sign, the user should be asked whether the physical height restricts any normal trafic including hgv. According to the answer we can tag:

1. No -> maxheight = default
2. Yes -> maxheight = below_default
3. Maybe (I can't estimate) -> maxheight = no_indications
4. I don't remember -> maxheight = no_sign

rugk commented:
HGV?

mmd-osm commented:
Hgv : Heavy goods vehicle (LKW in deutsch)

rugk commented:
Ah, "truck".😌

ENT8R commented:
> I don't remember -> maxheight = no_sign

Why should a surveyor who sees the bridge/tunnel in "real life" should not remember if he is standing right before it...?

matkoniecz commented:
I would recommend excluding vehicle=private|no ways.

matkoniecz commented:
"This height looks implausible" message hides value entered by user, maybe it should be somehow displayed in a message?

matkoniecz commented:
I would recommend excluding also access=private|no ways.

matkoniecz commented:
>  If there is no sign, should we tag it as maxheight=default (used about 10K times) or maxheight=no_sign as proposed by the wiki but only used 14 times according to taginfo...

I would use no_sign as it is more clear what is really mapped.

matkoniecz commented:
> should automatically a trailing .0 be added by SC?

I see no justification for that - why it should be done?

matkoniecz commented:
> I would not, because this decreases the amount of solveable quest significantly

Is it removing unwanted/pointless quests? If yes then I think that significant decrease makes it an even better idea. See cases like https://www.openstreetmap.org/way/518140477#map=19/50.07558/19.93155 or https://www.openstreetmap.org/way/295422615#map=19/50.05658/19.94075

What is the point of mapping maxheight restrictions for places inaccessible to general public? Owners of yard are not going to check maps/routing engines to check whatever they can access their own property.

ENT8R commented:
> Owners of yard are not going to check maps/routing engines to check whatever they can access their own property.

Well, that's true, but maybe there are some underground parkings which are not accessible by the general public but if you rent a parking space you can go in there... So there is not a single user but some more people who might be interested in using that data...

matkoniecz commented:
So maybe keep 

> node['amenity'='parking_entrance']['parking'~'^(underground|multi-storey)$']" +  QUERY_RESTRICTIONS + ";" +

unchanged and add access filters for

> way['highway'~'^(" + ROADS + ")$']['tunnel'~'^(yes|building_passage)$']" +  QUERY_RESTRICTIONS + ";" +

?

This should handle both cases.

ENT8R commented:
Good idea! I will do this.

ENT8R commented:
> I would recommend excluding vehicle=private|no ways.
I would recommend excluding also access=private|no ways.

I did this now, but at least in my city, this means that about 160 quests have been removed...

slhh commented:
> Don't think that this is necessary... The user will either remember the value he entered some moments ago or can simply press Cancel/I will check...

In case of a typo the user will likely remember the intended value, but not the typed value.

ENT8R commented:
> In case of a typo the user will likely remember the intended value, but not the typed value.

But isn't this the same with the max speed quest too?

westnordost commented:
Let's not put effort into that.

slhh commented:
>> I would recommend excluding vehicle=private|no ways.
>> I would recommend excluding also access=private|no ways.
> 
> I did this now, but at least in my city, this means that about 160 quests have been removed...

I see some potential reasons not to exclude:

1. There might be access key combinations like:
access=no + motorcar=yes + motorcycle=yes (e.g. for a motorway)
2. access=private might be incorrect
3. The quest are still valid with regards to data completeness, and seem to be easy to solve for walking users. In case of quests intended for users going by bus or car (hopefully not the driver), it is better to exclude.
4. Dependent on the region, the quests for public roads might be a bit demotivating, because finding a restriction is rare.

ENT8R commented:
I totally agree. And even if it is private, I don't think that this would be unnecessary data... Are there other opinions on that?

westnordost commented:
I also think including private is fine.

matkoniecz commented:
@ENT8R Is it now ready for testing? It seems that https://github.com/westnordost/StreetComplete/pull/960#discussion_r173544039 and https://github.com/westnordost/StreetComplete/pull/960#discussion_r173501214 are still waiting.

ENT8R commented:
> Is it now ready for testing?

Besides the comments you mentioned I need to find a better solution for the layout of countries with imperial unit (besides the US). I'm sorry, but I don't think that you need to test something right now. This makes IMO only sense if I fixed everything... I did not forget this PR but I had some other things to do... 😬

westnordost commented:
There are conflicts here now because I changed the system how the country specific resources are loaded. Perhaps I will take over this quest, but in any case it will not fit into v5 anymore.

matkoniecz commented:
> I also think including private is fine.

At least in my region there are many, many private service roads going through buildings like

![img_20180407_172436](https://user-images.githubusercontent.com/899988/40279289-217834d8-5c40-11e8-80a6-9a6282262fb9.jpg)

I think that tagging maxheight for all of them is not really useful (especially as it would be nearly universally "it is not signed").

Maybe at least exclude access=private + highway=service tunnels?

ENT8R commented:
> Maybe at least exclude access=private + highway=service

I think that would work. I'm adding this right now.

ENT8R commented:
> Some countries follow the [US standard (MUTCD)](https://en.wikipedia.org/wiki/Comparison_of_MUTCD-influenced_traffic_signs). In the linked page, search for "Height restriction ahead" to see how the sign look there. So, the layout should look like that.

But there are also signs for "Maximum height" which look often completely different... I would tend to use the signs for "Maximum height" because this is actually what this quest is about.

For example:

| Country | Height restriction ahead  | Maximum height |
| ------------- | ------------- | ------------- |
| Canada | <img src="https://upload.wikimedia.org/wikipedia/commons/f/f8/Qu%C3%A9bec_D-190-1.svg" width="75px">  | <img src="https://upload.wikimedia.org/wikipedia/commons/e/ea/Qu%C3%A9bec_D-190-2.svg" width="75px">  |
| USA | <img src="https://upload.wikimedia.org/wikipedia/commons/6/6a/MUTCD_W12-2.svg" width="75px">  | <img src="https://upload.wikimedia.org/wikipedia/commons/b/b3/MUTCD_W12-2a.svg" width="100px">  |
| Australia | <img src="https://upload.wikimedia.org/wikipedia/commons/c/cb/Australia_road_sign_W4-8.svg" width="75px">  | <img src="https://upload.wikimedia.org/wikipedia/commons/b/b0/Australia_R6-12.svg" width="100px">  |

westnordost commented:
Where is the difference between "Height restriction" and "Maximum height"?

ENT8R commented:
> Where is the difference between "Height restriction" and "Maximum height"?

No, the difference is that the left column shows signs which are only a "Vorankündigung" because it says [..] **ahead** and the right column shows the actual sign which is probably what we want...

westnordost commented:
Ah, I see

westnordost commented:
Two notes:
1. Apparently the sign in UK looks like this: https://en.wikipedia.org/wiki/File:UK_traffic_sign_629.2A.svg (imperial units + vienna convention-style sign). If *every* max height sign in UK has both units on the sign then I guess we (you) don't need to go through the trouble and change the layout to somehow include both. But if some (older) signs may still show only feet, then something must be done implementation-wise. Perhaps ask the British community. On the wikipedia article, it reads: 
> "All European countries use the metric system (distances in kilometres or metres; speeds in kilometres per hour; heights, widths and lengths in metres; weights in tonnes) with the exception of the United Kingdom, where distances, speeds, heights and widths are still indicated in imperial measurements (miles or yards; miles per hour; feet and inches), with an optional indication in metres on some limit restriction signs. Weight limits are shown in metric tonnes with a lower case "t", although some older signs still use the imperial long ton indicated by either an upper case "T" or simply the word "ton".
2. if the list of `heightUnit`s coincides with that of `speedUnit` (and `weightUnit`, for bridges), they should be unified into `distanceUnit` or even `measurementSystem` (metric vs imperial). I see that the list in `heightUnit` is shorter than `speedUnit`, perhaps you should do a little research on the diff.

ENT8R commented:
> If every max height sign in UK has both units on the sign 

From a quick look at https://wiki.openstreetmap.org/wiki/Key:maxheight#Photos (there are only photos from the UK) it seems that there are signs with both units but still some signs with only the imperial unit...

westnordost commented:
It'd be nice to have this in v6, but I have no time to help with this here (all hands full with building type / maxspeed). In case you can solve everything by yourself and it is ready for review, say the word.

ENT8R commented:
I could only continue to work on this at the end of the next week... But I think that the issues remaining are pretty easy to solve (so I can do it then).

ENT8R commented:
I think I solved now all remaining issues (except one new issue but I will elaborate this later) and this PR is ready for another review...

Here is a small overview of how the quest will look like in the different countries:

| Country | Quest | Country | Quest |
| ---- | --- | ---- | --- |
| Australia | <img src="https://user-images.githubusercontent.com/25306497/42046583-79a697ee-7afe-11e8-98cf-b867d6abfe29.png" width="200px"> | Canada | <img src="https://user-images.githubusercontent.com/25306497/42046642-9c338614-7afe-11e8-9b54-53c5adacfe19.png" width="200px"> |
| USA | <img src="https://user-images.githubusercontent.com/25306497/42046657-a455aa34-7afe-11e8-89fd-53152126c204.png" width="200px"> | Great Britain | <img src="https://user-images.githubusercontent.com/25306497/42046963-51af58ce-7aff-11e8-9a57-c4a20f936f06.png" width="200px"> |
| Europe and all other countries  | <img src="https://user-images.githubusercontent.com/25306497/42046955-4f1f37a0-7aff-11e8-96d2-32dd665ee55b.png" width="200px"> |

westnordost commented on app/src/main/java/de/westnordost/streetcomplete/quests/max_height/AddMaxHeightForm.java:
```diff
@@ -0,0 +1,262 @@
+
+		for (int i = 0; i < units.size(); i++) {
+			String unit = units.get(i);
+			if (unit.equals("metric"))
```
Shouldn't you use the constants here?

ENT8R replied:
Well, but they don't exist anymore if I'm using a enum there 😄

westnordost commented on app/src/main/java/de/westnordost/streetcomplete/quests/max_height/AddMaxHeightForm.java:
```diff
@@ -0,0 +1,262 @@
+			String input = heightInput.getText().toString();
+			if (!input.isEmpty())
+			{
+				if (input.contains("."))
```
Wouldn't that be a comma for German language (and whatnot for others)?

You should rather parse the string into a number (using the user's locale?) and then split up the number, if necessary. I don't really get why you do this for metric heights as well though. Can't you just have two constructors `Height(float meters)` and `Height(int feet, int inch)`?

westnordost commented:
Alright, reviewed. There are still some open points though. This will not get into v6, since even if this was ready now, it would still need translation, so no need to hurry.

ENT8R commented on app/src/main/assets/country_metadata/GB.yml:
```diff
@@ -4,10 +4,10 @@ isAdvisorySpeedLimitKnown: true
   sweet_pepper]
 popularReligions: [christian, muslim, jewish, hindu, sikh]
 popularSports: [soccer, tennis, bowls, cricket, golf, basketball, rugby]
-speedUnit: [mph]
```
You rather mean that maxspeed signs are always in mph, right? And AFAIK heights and lengths are (at least on the traffic signs) also mostly in imperial units...

westnordost replied:
Hmm, ok

ENT8R replied:
Wikipedia says here: https://en.wikipedia.org/wiki/Road_signs_in_the_United_Kingdom#Units_of_measurement
> The United Kingdom is the only European country to use mostly imperial units rather than SI units on road signs for distance measurements and speed limits, and one of few countries in the world [..] to do so. Vehicle weight limits are signed only in metric (TSRGD 1981), and metric units may be used in addition to imperial ones for height, width and length restrictions.

Just to give a source...

rory replied:
Can confirm. Speed limits in UK are always mph.

ENT8R commented on app/src/main/assets/country_metadata/LR.yml:
```diff
@@ -1,4 +1,5 @@
 # Do not edit. Source files are in /res/country_metadata
+measurementSystem: [imperial, metric]
```
Yes, Liberia was not included at all...

ENT8R commented on app/src/main/assets/country_metadata/PW.yml:
```diff
@@ -1,3 +1,4 @@
 # Do not edit. Source files are in /res/country_metadata
+measurementSystem: [imperial, metric]
```
I added [Palau](https://en.wikipedia.org/wiki/Palau) because it is a part of [Micronesia](https://en.wikipedia.org/wiki/Micronesia) (FM) which uses the imperial system because it is/was a territory of the US

ENT8R commented on app/src/main/assets/country_metadata/MH.yml:
```diff
@@ -1,4 +1,5 @@
 # Do not edit. Source files are in /res/country_metadata
+measurementSystem: [imperial, metric]
```
The same reason why I added Palau (PW): because the [Marshall Islands](https://en.wikipedia.org/wiki/Marshall_Islands) were a territory of the US but gained independence about 30 years ago but they are still in a conversion from imperial to metric units and traffic signs...

ENT8R commented on app/src/main/java/de/westnordost/streetcomplete/quests/max_height/Height.java:
```diff
@@ -0,0 +1,56 @@
+
+	public double toDouble()
+	{
+		return Double.parseDouble(integerPart + "." + fractionalPart);
```
The `toDouble()` method is just used for calculations (e.g. the "user selected an unrealistic height" method). The final value for OSM is retrieved with the `toString()` method...

westnordost replied:
then a method like
```
double getInMeters()
```
would be better, which already includes the conversion if necessary.

ENT8R commented on app/src/main/res/drawable-mcc240/background_maxheight_sign.xml:
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- Sweden SE -->
+<!-- TODO: this and the other two country-dependent files don't work -->
```
> Actually refactor `getCurrentCountryResources` to `getResources` (override)

But this can't be done because `getResources` is `final` and thus can't be overridden... Or am I missing something?
https://developer.android.com/reference/android/content/Context#getResources()

westnordost replied:
Which means
```
LayoutInflater.from(new ContextWrapper(getContext())
{
	@Override public Resources getResources()
	{
		return AbstractQuestAnswerFragment.this.getCurrentCountryResources();
	}
});
```

westnordost commented on app/src/main/res/layout/quest_maxheight_australia.xml:
```diff
@@ -0,0 +1,70 @@
+        android:layout_centerVertical="true"
+        android:layout_marginEnd="20dp"
+        android:background="#0fff"
+        android:visibility="visible" />
```
So you either have to include a `feetInputSign` here as well, or remove the `heightUnitSelect` here and in the class deal with that `heightUnitSelect` is null. I suggest you do whatever produces less code/layout (complexity).

westnordost commented on app/src/main/res/values/strings.xml:
```diff
@@ -605,4 +605,14 @@ Otherwise, you can download another keyboard in the app store. Popular keyboards
+    <string name="quest_maxheight_tunnel_title">What is the height limit of this tunnel?</string>
+    <string name="quest_maxheight_answer_noSign">There is no sign</string>
+    <string name="quest_maxheight_answer_noSign_question">Can you give some more information?</string>
+    <string name="quest_maxheight_answer_restrictsTraffic">The physical height restricts the general traffic</string>
```
So my thinking is: Why not ask the user to measure it? Or perhaps taking a photo would be posible (create a note).
I suggest the following:

> If you can instead measure the height, please do! Otherwise, is there enough space for even big trucks to pass?
> You can also leave a note and attach a photo to it.
> **Leave note | No | Yes**

And never tag any of `no_indications/unsigned/none/no_sign`, because that information is pretty worthless.

ENT8R replied:
Should there be a new dialog for measuring the height? And the tag for that purpose would be something like `maxheight:estimated`? Or should we still use `maxheight`?

ENT8R replied:
I would use an extra tag like `maxheight:estimated` because [`maxheight`](https://wiki.openstreetmap.org/wiki/Key:maxheight) is...

> The **legal** maximum height

...and nothing estimated. But maybe we would have to create a new wiki page for this tag because according to [taginfo](https://taginfo.openstreetmap.org/keys/maxheight%3Aestimated) this tag is currently used 0 times 😄
Maybe we could just change/add a hint like it is done for e.g. the vegetarian quest if the user wants to estimate the height but keep the layout as it is...

westnordost replied:
Though, an estimated height is, beyond stating that "it is high enough that anything can pass" next to useless. Basically, an estimated height can and should only be used as a marker for other people to supply the proper height.
So, perhaps it is the best to let the user just make the estimation whether it is **easily** (adapt wording) high enough for any vehicle to pass through or not. Only if the user had a measurement tape, it could be possible for him to add the precise height (and indeed that would be valuable information).
So, if the user answered no, mark it as, dunno, `no_sign` or something and this tag along with the others (`no_indications, unsigned, ...`) can be used as a marker for other surveyors that **do** bring a measurement tape to solve it. Maybe this could be a second "advanced" (default disabled) quest.

ENT8R commented:
I have currently no access to a computer, so it might take some time until I can finish this here...

ENT8R commented:
But thank you for your more extensive comments, maybe I should have told you earlier that I need some more time.

westnordost commented:
I [found evidence](http://agc.gov.ms/wp-content/uploads/2011/10/Road-Traffic-Act2.pdf) that Montserrat (MS) uses miles per hour. (Not that surprising, it is a British Overseas Territory)

westnordost commented:
Solved "Fix custom signs for Sweden, Finland etc. do not show up correctly" and some minor this-and-that.

ENT8R commented:
Just a small summary of what I did so far/what I found out:

- point 1 and 2 of your list should be fixed now. (I refactored the height class a little bit)
- the links in point 4 don't seem to work anymore... Is this fixed now?
- during testing in several countries I found a pretty worse bug which made point 3 even more worse: now there's (at least on my phone) always the standard input sign which is shown throughout Europe even in countries like Canada which should have their own layout. See for example this screenshot from Canada:
<img alt="canada" src="https://user-images.githubusercontent.com/25306497/43705911-2c021d86-9964-11e8-9938-c767a02724eb.png" width="150px">
Furthermore all signs in Finland/Iceland/... which should have a yellow background also look like this:
<img alt="finland" src="https://user-images.githubusercontent.com/25306497/43706013-78b772fc-9964-11e8-943d-49a64b97dbd5.png" width="150px">
I also downloaded the maxspeed quest for Canada and there everything works fine... It just seems that something with this quest or layout would be wrong...

westnordost commented:
Err I meant point 3.

It came to my mind that it would also be possible, and perhaps not that hard after all, to measure a height by making use of the accelerometer (and camera) instead of with a measure tape:
With this sensor, it is possible to calculate the angle in which the phone is held relative to the ground. Given that the height in which the phone is held is known (must be input manually in "calibration") and the user is given the instruction to 1. point the phone at the lower edge and 2. then at the upper edge of what he wants to measure, the height of the object can be calculated from these two angles and the mentioned height of the held phone.

![untitled](https://user-images.githubusercontent.com/4661658/43805210-2a5cf212-9a9f-11e8-83db-bdd7cf7dc671.png)

Though, while the calculation is easy, the UI, simplicity and clarity of use, is key for it being successful. Google Play is full with half-assed measurement utilities, no idea if there is a good one amongst the sea of mediocrity.

ENT8R commented:
> I guess a good UI would involve a AR-like line being shown on the camera

Do you mean something like a virtual horizon? I don't think that this is necessary for our use case. Isn't a simple cross hair enough?
I already found some apps and so far they seem to work pretty good. In my tests they always delivered the exact result.
Github is also full of those apps: https://github.com/search?l=Java&q=measure+height&type=Repositories :smile:

rugk commented:
So we may get some part of https://github.com/westnordost/StreetComplete/issues/879? 😄 

Or we just let [the user jump](https://github.com/julianeberle23/MaxJumpHeight)... 😉 (just kidding, sry.)

goldfndr commented on app/src/main/java/de/westnordost/streetcomplete/quests/max_height/Height.java:
```diff
@@ -34,24 +41,43 @@ public Unit getUnit()
+		}
+		else
+		{
+			return feetToMeter(Double.parseDouble(feet + "." + inches));
```
How about something like:
`return ((Integer.parseInt(feet) * 12 + Integer.parseInt(inches)) * 0.0254);`

This would do three things:

* Correctly interpret inches as a division of 12 instead of 10.
* Correctly return double precision (0.0254 is exact, 3.2808 is approximate and barely enough for float let alone double).
* Perform a multiply rather than a divide (less expensive).

(Note: untested; I don't know if more text ("Double") is required. I also don't know why Double is preferred over Float; is Double somehow quicker than Float?)

ENT8R commented:
Alright... I experimented around in the last days and I think I implemented what you proposed. From my point of view the UI is simple to use but I would be happy to receive any feedback.
During my tests some things came to my mind which I added as TODOs in the code:
- [should we use `maxheight:estimated=*` or `maxheight:triangulated=*` or something completely different for the measured height?](https://github.com/westnordost/StreetComplete/blob/545cb655665c5d6ff40b05cbb767ef5c86c407ad/app/src/main/java/de/westnordost/streetcomplete/quests/max_height/AddMaxHeight.java#L88)
- right now the app asks the user at the first use to enter the height he is holding the phone at and all other calculations are based on that value. The problem with this method might be that the user does not measure the exact height but rather makes a guess of the value. The other way would be to ask the user for his current height (this should be easier to answer; everybody should know their own height 😄), but we would have to subtract some centimeters afterwards (maybe something between 20cm to 30cm?) which is inaccurate too. (some people tend to hold their phone higher, some lower...) Are there any opinions on that? ([source](https://github.com/westnordost/StreetComplete/blob/545cb655665c5d6ff40b05cbb767ef5c86c407ad/app/src/main/java/de/westnordost/streetcomplete/quests/max_height/measure/MeasureCameraFragment.java#L258))
- currently the app checks everytime the user opens the "other answer menu" whether the device has a camera which can be used (this might be resource intense) and I even thought that the menu of this quest would open slower than other quests. ([source](https://github.com/westnordost/StreetComplete/blob/545cb655665c5d6ff40b05cbb767ef5c86c407ad/app/src/main/java/de/westnordost/streetcomplete/quests/max_height/AddMaxHeightForm.java#L122))
- I also didn't remove the dialog which was there before. This is only shown if the device does have no camera. ([source](https://github.com/westnordost/StreetComplete/blob/545cb655665c5d6ff40b05cbb767ef5c86c407ad/app/src/main/java/de/westnordost/streetcomplete/quests/max_height/AddMaxHeightForm.java#L142)) Are there many users/devices which don't have a camera at all and for which this dialog would still make sense or can I remove this completely?

The first two times the user uses this method to calculate the height there is a dialog with a pretty long instruction (maybe someone can proofread or even shorten it a little bit...) and two tooltips which instruct the user what to do next. (can also be seen on the screenshots below)
I am not satisfied yet with the help dialog... I tried to add a small illustration where the user can see what he has to do but I'm really bad at making pictograms 😄: (it looks like he's bleeding out of his eyes :laughing:)
![measure_height_help](https://user-images.githubusercontent.com/25306497/43991145-9c09da64-9d66-11e8-98f1-7be8d91439e1.png)
Do you think it would be helpful to display something like that? I just added it as a placeholder (so it can be removed pretty easy)

Actually the user only has to press two buttons after each other and rotate the phone in between to get a proper result. (I could not think of anything simplier...)
During the tests on my six-year-old Samsung S3 the average difference to the actual value was only at about 10-15 centimeters. If I would have measured the exact height of where I'm holding my phone at, t might have been even more precise. I also can imagine that newer and better phones have better sensors, so the deviation would be even smaller...

And here are some screenshots (the background looks so weird because I took the screenshots with my emulator which does not have a camera...):
<img src="https://user-images.githubusercontent.com/25306497/43991236-cbc03896-9d68-11e8-8076-66e39d64d7e2.png" width="300px"> <img src="https://user-images.githubusercontent.com/25306497/43991237-ce952842-9d68-11e8-8757-81339a9490dd.png" width="300px">
<img src="https://user-images.githubusercontent.com/25306497/43991238-d28b61f0-9d68-11e8-9ddc-6ffbab92c1f6.png" width="300px"> <img src="https://user-images.githubusercontent.com/25306497/43991239-d4ac280c-9d68-11e8-83db-27764c25a53e.png" width="300px">

rory commented on app/src/main/res/values/strings.xml:
```diff
@@ -619,6 +619,39 @@ Otherwise, you can download another keyboard in the app store. Popular keyboards
+        Make sure that you are holding the camera in the same position and height during the whole process.
+        You are only allowed to tilt the phone. Do not move while measuring the height because this might result in wrong values.
+        It is also important that you can see the whole object while measuring the height.
+        If the height is about 4 meter, you should stand at least 30 meter in front of the object.
```
> If the height is about 4 meter, you should stand at least 30 meter

Add an "s" (plural). "meter" is US English. "metre" is everywhere else.

ENT8R replied:
Well, but AFAIK the strings for American English are in the default folder at https://github.com/westnordost/StreetComplete/tree/master/app/src/main/res/values
and the ones for Great Britain are at https://github.com/westnordost/StreetComplete/tree/master/app/src/main/res/values-en-rGB
The reason why there are two files for the English language can be found here: #914

rory commented on app/src/main/res/values/strings.xml:
```diff
@@ -619,6 +619,39 @@ Otherwise, you can download another keyboard in the app store. Popular keyboards
+    </string>
+    <string name="height_measure_help_tips">
+        Make sure that you are holding the camera in the same position and height during the whole process.
+        You are only allowed to tilt the phone. Do not move while measuring the height because this might result in wrong values.
```
> You are only allowed to tilt the phone. Do not move while measuring the height

This wording is subpar. "move" in English doesn't just mean your changing lat/lon, it also means your arms and head. "Don't move" is like "freeze", where you can't move your arms or hands. "Tile the phone and don't move" is contradictory.

Perhaps "Stay in the same place"?

ENT8R replied:
Yes, that sounds better! Thank you!

rugk commented:
> right now the app asks the user at the first use to enter the height he is holding the phone at and all other calculations are based on that value. The problem with this method might be that the user does not measure the exact height but rather makes a guess of the value

Why not ask the user to lay the phone down at the bottom once and take it up again and hold in the "usual" height. I guess, with a good calculation/motion sensor, that would be a good way to "calibrate" the height. That would be a kinda silly/funny "ceremonial", but IMHO the best. And no, I would not assume anyone knows theirs size… and people may hold the phone at different heights. (actually the last thing could vary depending on what size you actually aim at – maybe you need to tell the user to hold it at eye-height or something like that) 
Maybe also: a) display the calculated size at the end anyway so the user can verify it (in case some sensor did not work. b) add the thing in the settings, so the user can see the currently calibrated height and re-do the process, if needed.

Also, of course, the calculated height at the end should also be displayed to the user before being added to OSM.
Maybe as for a general UI, just keep this slide-in quest UI and just add a button "measure with phone". It should not be hidden under "Other answers", at least.

And I know this is a first prototype, but of course you have to convert the text into bullet points or so. (simple Do's and Don'ts or something like that), so users can see on one glance what to do and what not.

rugk commented:
Another way you could do it is to ask the user to set up a ruler in front of them and do the process in the same way, i.e. mark top and bottom. They can then enter the size of the ruler.
This might be used as a verification step after the calibration, or you somehow include it in the calibration. Or something like that.

rugk commented:
> Maybe it makes more sense to have the camera-height-measurement happen in an own quest anyway

Great idea, because in such a case, you can move this conditional "device has a camera" (and maybe "has proper sensors") check into the check whether a quest is disabled. This can then be seen in the quest list by the user and they see a reason why it is disabled.

ENT8R commented:
Thank you for your feedback! I removed the whole height measurement thing and added the dialog instead which asks now: "Is it high enough for big trucks (higher than 4.5 meters) to pass through?" and in case of yes `maxheight=default` is added and in case of no, `maxheight=below_default` is added. I also downloaded some quests in Finland and Iceland and it still does not seem to work...
The idea to have this height measurement thing as a new quest is indeed much better than just another answer of this quest.

westnordost commented:
whoops, did a rebase-merge instead of a normal merge. In case there will be problems when merging back to master, I will fix them.

westnordost commented:
Okay I found it. It is not easily solvable, as Google did not anticipate I guess that one would like to work with modified Resources:
For reading the attributes during layout inflation, not directly the context's Resources are used, but the Context's Theme (`getTheme`). My `ContextWrapper` implementation in `AbstractQuestAnswerFragment` only overrides `getResources()`. While `getTheme` can also be overridden, the method's return type `Resources.Theme` is final, so I cannot create something like a `ThemeWrapper` that instead uses the overridden resources.
This means that resources specified in XML (backgrounds, strings, images, ...) will never be resolved to the current-country resources.

The workaround then is to manually set the resources again programmatically after inflation.

westnordost commented:
Also, in regards to the earlier problem I noticed: When you write a dot instead of a comma on a non-English locale, the app crashes
```
Caused by: java.text.ParseException: Unparseable number: "14.5" (at offset 4)
                      at java.text.NumberFormat.parse(NumberFormat.java:557)
                      at de.westnordost.streetcomplete.quests.max_height.AddMaxHeightForm.getHeightFromInput(AddMaxHeightForm.java:192)
```

goldfndr commented:
Although the measure_height_help image is gone from this PR, I have some feedback on it for when it next appears.

* The object should have a flat top and bottom, to encourage accuracy.
* The person and object should be at the same level.
* A flat elevation line should extend from the object to the person, to emphasize that a slope gives less accuracy.

ENT8R commented:
Thank you very much for finishing this! :+1: :tada: I was really busy in the last days...

westnordost commented:
You're welcome! But also, thank you of course for preparing this PR :-)
